active/0021-transfer-files-over-mqtt.md

```diff
@@ -0,0 +1,175 @@
+* Separate data transfer and file transfer channels to achieve non-blocking transmission
+  * Channel priority setting, specify the priority of different channels according to customer requirements
+  * Channels can cooperate with each other to achieve overall flow control
+  * Multiple channels can be specified for file transfer at the same time to increase file transfer parallelism
+* Efficient transmission support in weak network
+  * File transfer in weak networks such as Telematics and mobile
+  * Automatic file chunking, block-based granularity breakpoint transfer (or file-level breakpoint transfer?)
+  * Switch to TCP if QUIC is not available
+
+## AWS IoT MQTT-based file delivery (reference design)
+
+As an example of existing implementation we can look at AWS IoT Core [which provides functionality](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery.html) to [deliver files to IoT devices](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt-based-file-delivery-in-devices.html):
+
+![AWS IoT MQTT-based file delivery Device --> Cloud](0021-assets/aws-mqtt-file-delivery.png)
+
+## Design
+
+In the first phase we are going to focus on limited scope:
+
+* device-to-cloud file transfer only
+* reusing existing TCP connection
+* push only, i.e. broker cannot request file transfer, but device can start sending files any time
+* device may not know total file size when starting the transfer
+
+### Requirements
+
+* Based on the MQTT protocol. Use MQTT messages to encapsulate file bodies
+* Transport layer TCP, TCP/TLS and QUIC (TCP fallback) multiplexing connections with other service topics​
+* High concurrency (multiple uploads in progress)
+* MQTT Client single/bi-directional transfer (active file upload, passive file download)
+* Fault tolerance: disconnected retransmission (sender side)
+* Fragmentation: (multi parts, fragments)
+* Multi-channel: multiple TCP connections, QUIC multi-streams
+* Cancellation: Both parties can pause or cancel the transfer, and an event notification is required when the transfer is complete. (How to notify? What is the target of the notification?)
+* Minimal code changes on Client side, SDK support available. (low code/no code)
+* Minimize data retransmission, and reduce network pressure and bandwidth usage.
+* RestAPI interface Query transfer status
+* EMQX local incomplete file cache
+* Client API should be easy to use: easy to migrate from existing applications (ftp, nfs, s3)
+* File body Encryption and decryption (with external key management)?
+
+### Out of scope
+
+* Non-reliable transmission is not supported
+* No support for one-to-many transfers
+* No distributed cache storage
+* Broker does not aware of interrupted transfer, It will be client-side implementation
+
+### Limitations
+
+* File transfer is PUSH mechanism but not support POLL, the receiving side cannot request to initiate the transfer
+decisions of the size of a slice cannot be as natural as streaming.
```

Comment:
Should this be a separate item?  Or is it indeed part of the previous phrase in the previous line?